File: src/topic_example/topic_listener/src/topic_listener.cc
/******************************************************************************
 * Copyright (c) 2022-2026 SEGAR. All Rights Reserved.
 * SPDX-License-Identifier: LicenseRef-Segar-Proprietary
 *
 * PROPRIETARY AND CONFIDENTIAL. See ./LICENSE
 * for license terms and restrictions.
 *****************************************************************************/

#include <memory>
#include <string>

#include "example/msg/String.hpp"

#include "segar/segar.h"

int main(int argc, char* argv[]) {
  RETURN_VAL_IF(!rti::segar::Init(argv[0]), EXIT_FAILURE);

  auto node = rti::segar::CreateNode("topic_listener");
  RETURN_VAL_IF(!node, EXIT_FAILURE);
  auto reader = node->CreateReader<example::msg::String>(
      "/topic/chatter",
      [](const auto& msg) { AINFO << "Received message: " << msg->data(); });
  RETURN_VAL_IF(!reader, EXIT_FAILURE);
  AINFO << "Waiting for messages...";
  rti::segar::WaitForShutdown();

  return EXIT_SUCCESS;
}

File: src_python/topic_example/topic_listener/src/topic_listener.py
#!/usr/bin/env python3

from example.msg import String
from segar.python import segar


def main():
    if not segar.Init("topic_listener"):
        return 1
    node = segar.CreateNode("topic_listener")
    if node is None:
        return 1

    def on_msg(msg):
        print(f"[topic_listener] recv: {msg.data}")

    reader = node.CreateReader("/topic/chatter", String, on_msg)
    if reader is None:
        print("[topic_listener] create reader failed")
        return 1

    print("[topic_listener] waiting for messages...")
    segar.WaitForShutdown()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
